sourmash/sbtmh.py

```diff
@@ -77,6 +80,207 @@ def data(self, new_data):
         self._data = new_data
 
 
+class LocalizedSBT(SBT):
+    """A Sequence Bloom Tree implementation which guarantees new leaves are plaaced in
+
+    The default node is a Bloom Filter (like the original implementation),
+    and the leaves are MinHash leaf class (in the sourmash.sbtmh.SigLeaf class)
+
+    Parameters
+    ----------
+    factory: Factory
+        Callable for generating new datastores for internal nodes.
+    d: int
+        Number of children for each internal node. Defaults to 2 (a binary tree)
+    storage: Storage, default: None
+        A Storage is any place where we can save and load data for the nodes.
+        If set to None, will use a FSStorage.
+
+    Notes
+    -----
+    We use two dicts to store the tree structure: One for the internal nodes,
+    and another for the leaves (datasets).
+    """
+
+    def __init__(self, factory, d=2, storage=None, track_abundance=False,
+                 do_containment=False):
+        if d != 2:
+            raise NotImplementedError("LocalizedSBT is only implemented for when the "
+                                      "number of children is 2, d=2")
+        SBT.__init__(self, factory=factory, d=d, storage=storage)
+        self.track_abundance = track_abundance
+        self.ignore_abundance = not self.track_abundance
+        self.do_containment = do_containment
+
+    def new_node_pos(self, node):
+        if not self._nodes:
+            self.next_node = 1
+            return 0
+
+        if not self._leaves:
+            self.next_node = 2
+            return 1
+        # Not an empty tree, can search
+
+        # TODO: There is probably a way better way to write this logic - @olgabot
```

Comment:
Right, I would like to suggest that the code below be refactored into smaller functions. It's already pretty clear (good job w/comments and formatting!!) but it's complex enough to warrant more :). I also want to place a marker that this code should also have a lot of associated tests, perhaps at the unit test level. We have a lot of functional tests on SBTs, so I'm reasonably sure the code works, but it would be good to get some edge cases and so on at the unit test level in the code.

Reply:
i think it would be cool to incorporate some factory methods to contain a fair amount of the if/else logic.